{
 "cells": [
  {
   "attachments": {},
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "---\n",
    "title: Crossref Reader\n",
    "description: |\n",
    "  This notebook shows how to read metadata from Crossref, using a Crossref DOI.\n",
    "---"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fetch the metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'regex._regex'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[1;32m/Users/mfenner/Documents/commonmeta-py/docs/readers/crossref_reader.ipynb Zelle 2\u001b[0m line \u001b[0;36m1\n\u001b[0;32m----> <a href='vscode-notebook-cell:/Users/mfenner/Documents/commonmeta-py/docs/readers/crossref_reader.ipynb#W2sZmlsZQ%3D%3D?line=0'>1</a>\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mcommonmeta\u001b[39;00m \u001b[39mimport\u001b[39;00m Metadata\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/mfenner/Documents/commonmeta-py/docs/readers/crossref_reader.ipynb#W2sZmlsZQ%3D%3D?line=2'>3</a>\u001b[0m \u001b[39m# Fetch metadata from a Crossref DOI\u001b[39;00m\n\u001b[1;32m      <a href='vscode-notebook-cell:/Users/mfenner/Documents/commonmeta-py/docs/readers/crossref_reader.ipynb#W2sZmlsZQ%3D%3D?line=3'>4</a>\u001b[0m string \u001b[39m=\u001b[39m \u001b[39m'\u001b[39m\u001b[39m10.7554/elife.01567\u001b[39m\u001b[39m'\u001b[39m\n",
      "File \u001b[0;32m~/Library/Python/3.11/lib/python/site-packages/commonmeta_py-0.6.0-py3.11.egg/commonmeta/__init__.py:17\u001b[0m\n\u001b[1;32m     14\u001b[0m __author__ \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mMartin Fenner\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m     15\u001b[0m __license__ \u001b[39m=\u001b[39m \u001b[39m\"\u001b[39m\u001b[39mMIT\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[0;32m---> 17\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mmetadata\u001b[39;00m \u001b[39mimport\u001b[39;00m Metadata\n",
      "File \u001b[0;32m~/Library/Python/3.11/lib/python/site-packages/commonmeta_py-0.6.0-py3.11.egg/commonmeta/metadata/__init__.py:3\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39m# -*- coding: utf-8 -*-\u001b[39;00m\n\u001b[0;32m----> 3\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mmetadata\u001b[39;00m \u001b[39mimport\u001b[39;00m Metadata\n",
      "File \u001b[0;32m~/Library/Python/3.11/lib/python/site-packages/commonmeta_py-0.6.0-py3.11.egg/commonmeta/metadata/metadata.py:8\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39myaml\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mxmltodict\u001b[39;00m\n\u001b[0;32m----> 8\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mreaders\u001b[39;00m \u001b[39mimport\u001b[39;00m (\n\u001b[1;32m      9\u001b[0m     get_crossref,\n\u001b[1;32m     10\u001b[0m     read_crossref,\n\u001b[1;32m     11\u001b[0m     get_datacite,\n\u001b[1;32m     12\u001b[0m     read_datacite,\n\u001b[1;32m     13\u001b[0m     read_datacite_xml,\n\u001b[1;32m     14\u001b[0m     get_crossref_xml,\n\u001b[1;32m     15\u001b[0m     read_crossref_xml,\n\u001b[1;32m     16\u001b[0m     get_schema_org,\n\u001b[1;32m     17\u001b[0m     read_schema_org,\n\u001b[1;32m     18\u001b[0m     get_codemeta,\n\u001b[1;32m     19\u001b[0m     read_citeproc,\n\u001b[1;32m     20\u001b[0m     read_codemeta,\n\u001b[1;32m     21\u001b[0m     get_cff,\n\u001b[1;32m     22\u001b[0m     read_cff\n\u001b[1;32m     23\u001b[0m )\n\u001b[1;32m     24\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mwriters\u001b[39;00m \u001b[39mimport\u001b[39;00m (\n\u001b[1;32m     25\u001b[0m     write_datacite,\n\u001b[1;32m     26\u001b[0m     write_bibtex,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     31\u001b[0m     write_commonmeta\n\u001b[1;32m     32\u001b[0m )\n\u001b[1;32m     33\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mutils\u001b[39;00m \u001b[39mimport\u001b[39;00m normalize_id, find_from_format\n",
      "File \u001b[0;32m~/Library/Python/3.11/lib/python/site-packages/commonmeta_py-0.6.0-py3.11.egg/commonmeta/readers/__init__.py:2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[39m\"\"\"Readers for different metadata formats\"\"\"\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mcrossref_reader\u001b[39;00m \u001b[39mimport\u001b[39;00m get_crossref, read_crossref\n\u001b[1;32m      3\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mdatacite_reader\u001b[39;00m \u001b[39mimport\u001b[39;00m get_datacite, read_datacite\n\u001b[1;32m      4\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mschema_org_reader\u001b[39;00m \u001b[39mimport\u001b[39;00m get_schema_org, read_schema_org\n",
      "File \u001b[0;32m~/Library/Python/3.11/lib/python/site-packages/commonmeta_py-0.6.0-py3.11.egg/commonmeta/readers/crossref_reader.py:16\u001b[0m\n\u001b[1;32m     14\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mbase_utils\u001b[39;00m \u001b[39mimport\u001b[39;00m wrap, compact, presence, sanitize\n\u001b[1;32m     15\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mauthor_utils\u001b[39;00m \u001b[39mimport\u001b[39;00m get_authors\n\u001b[0;32m---> 16\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mdate_utils\u001b[39;00m \u001b[39mimport\u001b[39;00m get_date_from_date_parts\n\u001b[1;32m     17\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mdoi_utils\u001b[39;00m \u001b[39mimport\u001b[39;00m doi_as_url, doi_from_url, get_doi_ra, crossref_api_url\n\u001b[1;32m     18\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mconstants\u001b[39;00m \u001b[39mimport\u001b[39;00m (\n\u001b[1;32m     19\u001b[0m     CR_TO_BIB_TRANSLATIONS,\n\u001b[1;32m     20\u001b[0m     CR_TO_CP_TRANSLATIONS,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     24\u001b[0m     Commonmeta,\n\u001b[1;32m     25\u001b[0m )\n",
      "File \u001b[0;32m~/Library/Python/3.11/lib/python/site-packages/commonmeta_py-0.6.0-py3.11.egg/commonmeta/date_utils.py:5\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mdatetime\u001b[39;00m \u001b[39mimport\u001b[39;00m datetime \u001b[39mas\u001b[39;00m dt\n\u001b[1;32m      4\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mtyping\u001b[39;00m \u001b[39mimport\u001b[39;00m Optional, Union\n\u001b[0;32m----> 5\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mdateparser\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mpydash\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mpy_\u001b[39;00m\n\u001b[1;32m      7\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mutils\u001b[39;00m \u001b[39mimport\u001b[39;00m wrap\n",
      "File \u001b[0;32m~/.pyenv/versions/3.9.16/lib/python3.9/site-packages/dateparser/__init__.py:3\u001b[0m\n\u001b[1;32m      1\u001b[0m __version__ \u001b[39m=\u001b[39m \u001b[39m'\u001b[39m\u001b[39m1.1.8\u001b[39m\u001b[39m'\u001b[39m\n\u001b[0;32m----> 3\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mdate\u001b[39;00m \u001b[39mimport\u001b[39;00m DateDataParser\n\u001b[1;32m      4\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mconf\u001b[39;00m \u001b[39mimport\u001b[39;00m apply_settings\n\u001b[1;32m      6\u001b[0m _default_parser \u001b[39m=\u001b[39m DateDataParser()\n",
      "File \u001b[0;32m~/.pyenv/versions/3.9.16/lib/python3.9/site-packages/dateparser/date.py:6\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mcollections\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mabc\u001b[39;00m \u001b[39mimport\u001b[39;00m Set\n\u001b[1;32m      4\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mdatetime\u001b[39;00m \u001b[39mimport\u001b[39;00m datetime, timedelta\n\u001b[0;32m----> 6\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mregex\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39mre\u001b[39;00m\n\u001b[1;32m      7\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mtzlocal\u001b[39;00m \u001b[39mimport\u001b[39;00m get_localzone\n\u001b[1;32m      8\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mdateutil\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39mrelativedelta\u001b[39;00m \u001b[39mimport\u001b[39;00m relativedelta\n",
      "File \u001b[0;32m~/.pyenv/versions/3.9.16/lib/python3.9/site-packages/regex/__init__.py:1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m\u001b[39mregex\u001b[39;00m \u001b[39mimport\u001b[39;00m \u001b[39m*\u001b[39m\n\u001b[1;32m      2\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39m.\u001b[39;00m \u001b[39mimport\u001b[39;00m regex\n\u001b[1;32m      3\u001b[0m __all__ \u001b[39m=\u001b[39m regex\u001b[39m.\u001b[39m__all__\n",
      "File \u001b[0;32m~/.pyenv/versions/3.9.16/lib/python3.9/site-packages/regex/regex.py:421\u001b[0m\n\u001b[1;32m    416\u001b[0m     \u001b[39mreturn\u001b[39;00m r\n\u001b[1;32m    418\u001b[0m \u001b[39m# --------------------------------------------------------------------\u001b[39;00m\n\u001b[1;32m    419\u001b[0m \u001b[39m# Internals.\u001b[39;00m\n\u001b[0;32m--> 421\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mregex\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39m_regex_core\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39m_regex_core\u001b[39;00m\n\u001b[1;32m    422\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mregex\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39m_regex\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39m_regex\u001b[39;00m\n\u001b[1;32m    423\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mthreading\u001b[39;00m \u001b[39mimport\u001b[39;00m RLock \u001b[39mas\u001b[39;00m _RLock\n",
      "File \u001b[0;32m~/.pyenv/versions/3.9.16/lib/python3.9/site-packages/regex/_regex_core.py:21\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39municodedata\u001b[39;00m\n\u001b[1;32m     19\u001b[0m \u001b[39mfrom\u001b[39;00m \u001b[39mcollections\u001b[39;00m \u001b[39mimport\u001b[39;00m defaultdict\n\u001b[0;32m---> 21\u001b[0m \u001b[39mimport\u001b[39;00m \u001b[39mregex\u001b[39;00m\u001b[39m.\u001b[39;00m\u001b[39m_regex\u001b[39;00m \u001b[39mas\u001b[39;00m \u001b[39m_regex\u001b[39;00m\n\u001b[1;32m     23\u001b[0m __all__ \u001b[39m=\u001b[39m [\u001b[39m\"\u001b[39m\u001b[39mA\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mASCII\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mB\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mBESTMATCH\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mD\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mDEBUG\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mE\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mENHANCEMATCH\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m     24\u001b[0m   \u001b[39m\"\u001b[39m\u001b[39mF\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mFULLCASE\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mI\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mIGNORECASE\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mL\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mLOCALE\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mM\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mMULTILINE\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mP\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m     25\u001b[0m   \u001b[39m\"\u001b[39m\u001b[39mPOSIX\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mR\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mREVERSE\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mS\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mDOTALL\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mT\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mTEMPLATE\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mU\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mUNICODE\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m     26\u001b[0m   \u001b[39m\"\u001b[39m\u001b[39mV0\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mVERSION0\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mV1\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mVERSION1\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mW\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mWORD\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mX\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mVERBOSE\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39merror\u001b[39m\u001b[39m\"\u001b[39m,\n\u001b[1;32m     27\u001b[0m   \u001b[39m\"\u001b[39m\u001b[39mScanner\u001b[39m\u001b[39m\"\u001b[39m, \u001b[39m\"\u001b[39m\u001b[39mRegexFlag\u001b[39m\u001b[39m\"\u001b[39m]\n\u001b[1;32m     29\u001b[0m \u001b[39m# The regex exception.\u001b[39;00m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'regex._regex'"
     ]
    }
   ],
   "source": [
    "from commonmeta import Metadata\n",
    "\n",
    "# Fetch metadata from a Crossref DOI\n",
    "string = '10.7554/elife.01567'\n",
    "metadata = Metadata(string)\n",
    "\n",
    "# Check that metadata was fetched successfully\n",
    "print(metadata.state)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Inspect the metadata\n",
    "\n",
    "The following metadata are required for each Crossref resource:\n",
    "\n",
    "* `id`: the persistent identifier of the resource, e.g. a DOI\n",
    "* `type`: the type of the resource in commonmeta format, e.g `JournalArticle`, `Book`, `BookChapter`, or `Dataset`.\n",
    "* `titles`: the title(s) of the resource\n",
    "* `creators`: the creator(s)/author(s) of the resource\n",
    "* `publisher`: the publisher of the resource\n",
    "* `publication_year`: the publication year of the resource\n",
    "\n",
    "In addition, there are plenty of optional metadata. They are converted into the internal commonmeta format. One example is the `licence` which is mapped to the Software Package Data Exchange ([SPDX](https://spdx.dev/)) vocabulary. For the publication from *eLife* we fetched earlier, this means a CC-BY license."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "    \"id\": \"https://doi.org/10.7554/elife.01567\",\n",
      "    \"type\": \"JournalArticle\",\n",
      "    \"url\": \"https://elifesciences.org/articles/01567\",\n",
      "    \"creators\": [\n",
      "        {\n",
      "            \"type\": \"Person\",\n",
      "            \"givenName\": \"Martial\",\n",
      "            \"familyName\": \"Sankar\",\n",
      "            \"affiliation\": [\n",
      "                {\n",
      "                    \"name\": \"Department of Plant Molecular Biology, University of Lausanne, Lausanne, Switzerland\"\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"type\": \"Person\",\n",
      "            \"givenName\": \"Kaisa\",\n",
      "            \"familyName\": \"Nieminen\",\n",
      "            \"affiliation\": [\n",
      "                {\n",
      "                    \"name\": \"Department of Plant Molecular Biology, University of Lausanne, Lausanne, Switzerland\"\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"type\": \"Person\",\n",
      "            \"givenName\": \"Laura\",\n",
      "            \"familyName\": \"Ragni\",\n",
      "            \"affiliation\": [\n",
      "                {\n",
      "                    \"name\": \"Department of Plant Molecular Biology, University of Lausanne, Lausanne, Switzerland\"\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"type\": \"Person\",\n",
      "            \"givenName\": \"Ioannis\",\n",
      "            \"familyName\": \"Xenarios\",\n",
      "            \"affiliation\": [\n",
      "                {\n",
      "                    \"name\": \"Vital-IT, Swiss Institute of Bioinformatics, Lausanne, Switzerland\"\n",
      "                }\n",
      "            ]\n",
      "        },\n",
      "        {\n",
      "            \"type\": \"Person\",\n",
      "            \"givenName\": \"Christian S\",\n",
      "            \"familyName\": \"Hardtke\",\n",
      "            \"affiliation\": [\n",
      "                {\n",
      "                    \"name\": \"Department of Plant Molecular Biology, University of Lausanne, Lausanne, Switzerland\"\n",
      "                }\n",
      "            ]\n",
      "        }\n",
      "    ],\n",
      "    \"titles\": [\n",
      "        {\n",
      "            \"title\": \"Automated quantitative histology reveals vascular morphodynamics during Arabidopsis hypocotyl secondary growth\"\n",
      "        }\n",
      "    ],\n",
      "    \"publisher\": {\n",
      "        \"id\": \"https://api.crossref.org/members/4374\",\n",
      "        \"name\": \"eLife Sciences Publications, Ltd\"\n",
      "    },\n",
      "    \"date\": {\n",
      "        \"published\": \"2014-02-11\",\n",
      "        \"updated\": \"2022-03-26T09:21:50Z\"\n",
      "    },\n",
      "    \"subjects\": [\n",
      "        {\n",
      "            \"subject\": \"General Immunology and Microbiology\"\n",
      "        },\n",
      "        {\n",
      "            \"subject\": \"General Biochemistry, Genetics and Molecular Biology\"\n",
      "        },\n",
      "        {\n",
      "            \"subject\": \"General Medicine\"\n",
      "        },\n",
      "        {\n",
      "            \"subject\": \"General Neuroscience\"\n",
      "        }\n",
      "    ],\n",
      "    \"language\": \"en\",\n",
      "    \"references\": [\n",
      "        {\n",
      "            \"key\": \"bib1\",\n",
      "            \"doi\": \"https://doi.org/10.1038/nature02100\",\n",
      "            \"creator\": \"Bonke\",\n",
      "            \"title\": \"APL regulates vascular tissue identity in Arabidopsis\",\n",
      "            \"publicationYear\": \"2003\",\n",
      "            \"volume\": \"426\",\n",
      "            \"firstPage\": \"181\",\n",
      "            \"containerTitle\": \"Nature\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib2\",\n",
      "            \"doi\": \"https://doi.org/10.1534/genetics.109.104976\",\n",
      "            \"creator\": \"Brenner\",\n",
      "            \"title\": \"In the beginning was the worm\",\n",
      "            \"publicationYear\": \"2009\",\n",
      "            \"volume\": \"182\",\n",
      "            \"firstPage\": \"413\",\n",
      "            \"containerTitle\": \"Genetics\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib3\",\n",
      "            \"doi\": \"https://doi.org/10.1034/j.1399-3054.2002.1140413.x\",\n",
      "            \"creator\": \"Chaffey\",\n",
      "            \"title\": \"Secondary xylem development in Arabidopsis: a model for wood formation\",\n",
      "            \"publicationYear\": \"2002\",\n",
      "            \"volume\": \"114\",\n",
      "            \"firstPage\": \"594\",\n",
      "            \"containerTitle\": \"Physiologia Plantarum\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib4\",\n",
      "            \"doi\": \"https://doi.org/10.1162/089976601750399335\",\n",
      "            \"creator\": \"Chang\",\n",
      "            \"title\": \"Training nu-support vector classifiers: theory and algorithms\",\n",
      "            \"publicationYear\": \"2001\",\n",
      "            \"volume\": \"13\",\n",
      "            \"firstPage\": \"2119\",\n",
      "            \"containerTitle\": \"Neural computation\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib5\",\n",
      "            \"doi\": \"https://doi.org/10.1007/bf00994018\",\n",
      "            \"creator\": \"Cortes\",\n",
      "            \"title\": \"Support-vector Networks\",\n",
      "            \"publicationYear\": \"1995\",\n",
      "            \"volume\": \"20\",\n",
      "            \"firstPage\": \"273\",\n",
      "            \"containerTitle\": \"Machine Learning\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib6\",\n",
      "            \"doi\": \"https://doi.org/10.1242/dev.119.1.71\",\n",
      "            \"creator\": \"Dolan\",\n",
      "            \"title\": \"Cellular organisation of the Arabidopsis thaliana root\",\n",
      "            \"publicationYear\": \"1993\",\n",
      "            \"volume\": \"119\",\n",
      "            \"firstPage\": \"71\",\n",
      "            \"containerTitle\": \"Development\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib7\",\n",
      "            \"doi\": \"https://doi.org/10.1016/j.semcdb.2009.09.009\",\n",
      "            \"creator\": \"Elo\",\n",
      "            \"title\": \"Stem cell function during plant vascular development\",\n",
      "            \"publicationYear\": \"2009\",\n",
      "            \"volume\": \"20\",\n",
      "            \"firstPage\": \"1097\",\n",
      "            \"containerTitle\": \"Seminars in Cell & Developmental Biology\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib8\",\n",
      "            \"doi\": \"https://doi.org/10.1242/dev.091314\",\n",
      "            \"creator\": \"Etchells\",\n",
      "            \"title\": \"WOX4 and WOX14 act downstream of the PXY receptor kinase to regulate plant vascular proliferation independently of any role in vascular organisation\",\n",
      "            \"publicationYear\": \"2013\",\n",
      "            \"volume\": \"140\",\n",
      "            \"firstPage\": \"2224\",\n",
      "            \"containerTitle\": \"Development\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib9\",\n",
      "            \"doi\": \"https://doi.org/10.1371/journal.pgen.1002997\",\n",
      "            \"creator\": \"Etchells\",\n",
      "            \"title\": \"Plant vascular cell division is maintained by an interaction between PXY and ethylene signalling\",\n",
      "            \"publicationYear\": \"2012\",\n",
      "            \"volume\": \"8\",\n",
      "            \"firstPage\": \"e1002997\",\n",
      "            \"containerTitle\": \"PLOS Genetics\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib10\",\n",
      "            \"doi\": \"https://doi.org/10.1038/msb.2010.25\",\n",
      "            \"creator\": \"Fuchs\",\n",
      "            \"title\": \"Clustering phenotype populations by genome-wide RNAi and multiparametric imaging\",\n",
      "            \"publicationYear\": \"2010\",\n",
      "            \"volume\": \"6\",\n",
      "            \"firstPage\": \"370\",\n",
      "            \"containerTitle\": \"Molecular Systems Biology\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib11\",\n",
      "            \"doi\": \"https://doi.org/10.1016/j.biosystems.2012.07.004\",\n",
      "            \"creator\": \"Granqvist\",\n",
      "            \"title\": \"BaSAR-A tool in R for frequency detection\",\n",
      "            \"publicationYear\": \"2012\",\n",
      "            \"volume\": \"110\",\n",
      "            \"firstPage\": \"60\",\n",
      "            \"containerTitle\": \"Bio Systems\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib12\",\n",
      "            \"doi\": \"https://doi.org/10.1016/j.pbi.2005.11.013\",\n",
      "            \"creator\": \"Groover\",\n",
      "            \"title\": \"Developmental mechanisms regulating secondary growth in woody plants\",\n",
      "            \"publicationYear\": \"2006\",\n",
      "            \"volume\": \"9\",\n",
      "            \"firstPage\": \"55\",\n",
      "            \"containerTitle\": \"Current Opinion in Plant Biology\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib13\",\n",
      "            \"doi\": \"https://doi.org/10.1105/tpc.110.076083\",\n",
      "            \"creator\": \"Hirakawa\",\n",
      "            \"title\": \"TDIF peptide signaling regulates vascular stem cell proliferation via the WOX4 homeobox gene in Arabidopsis\",\n",
      "            \"publicationYear\": \"2010\",\n",
      "            \"volume\": \"22\",\n",
      "            \"firstPage\": \"2618\",\n",
      "            \"containerTitle\": \"Plant Cell\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib14\",\n",
      "            \"doi\": \"https://doi.org/10.1073/pnas.0808444105\",\n",
      "            \"creator\": \"Hirakawa\",\n",
      "            \"title\": \"Non-cell-autonomous control of vascular stem cell fate by a CLE peptide/receptor system\",\n",
      "            \"publicationYear\": \"2008\",\n",
      "            \"volume\": \"105\",\n",
      "            \"firstPage\": \"15208\",\n",
      "            \"containerTitle\": \"Proceedings of the National Academy of Sciences of the United States of America\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib15\",\n",
      "            \"doi\": \"https://doi.org/10.1016/0092-8674(89)90900-8\",\n",
      "            \"creator\": \"Meyerowitz\",\n",
      "            \"title\": \"Arabidopsis, a useful weed\",\n",
      "            \"publicationYear\": \"1989\",\n",
      "            \"volume\": \"56\",\n",
      "            \"firstPage\": \"263\",\n",
      "            \"containerTitle\": \"Cell\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib16\",\n",
      "            \"doi\": \"https://doi.org/10.1126/science.1066609\",\n",
      "            \"creator\": \"Meyerowitz\",\n",
      "            \"title\": \"Plants compared to animals: the broadest comparative study of development\",\n",
      "            \"publicationYear\": \"2002\",\n",
      "            \"volume\": \"295\",\n",
      "            \"firstPage\": \"1482\",\n",
      "            \"containerTitle\": \"Science\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib17\",\n",
      "            \"doi\": \"https://doi.org/10.1104/pp.104.040212\",\n",
      "            \"creator\": \"Nieminen\",\n",
      "            \"title\": \"A weed for wood? Arabidopsis as a genetic model for xylem development\",\n",
      "            \"publicationYear\": \"2004\",\n",
      "            \"volume\": \"135\",\n",
      "            \"firstPage\": \"653\",\n",
      "            \"containerTitle\": \"Plant Physiol\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib18\",\n",
      "            \"doi\": \"https://doi.org/10.1038/nbt1206-1565\",\n",
      "            \"creator\": \"Noble\",\n",
      "            \"title\": \"What is a support vector machine?\",\n",
      "            \"publicationYear\": \"2006\",\n",
      "            \"volume\": \"24\",\n",
      "            \"firstPage\": \"1565\",\n",
      "            \"containerTitle\": \"Nature Biotechnology\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib19\",\n",
      "            \"doi\": \"https://doi.org/10.1073/pnas.77.3.1516\",\n",
      "            \"creator\": \"Olson\",\n",
      "            \"title\": \"Classification of cultured mammalian cells by shape analysis and pattern recognition\",\n",
      "            \"publicationYear\": \"1980\",\n",
      "            \"volume\": \"77\",\n",
      "            \"firstPage\": \"1516\",\n",
      "            \"containerTitle\": \"Proceedings of the National Academy of Sciences of the United States of America\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib20\",\n",
      "            \"doi\": \"https://doi.org/10.1093/bioinformatics/btq046\",\n",
      "            \"creator\": \"Pau\",\n",
      "            \"title\": \"EBImage\\u2013an R package for image processing with applications to cellular phenotypes\",\n",
      "            \"publicationYear\": \"2010\",\n",
      "            \"volume\": \"26\",\n",
      "            \"firstPage\": \"979\",\n",
      "            \"containerTitle\": \"Bioinformatics\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib21\",\n",
      "            \"doi\": \"https://doi.org/10.1105/tpc.111.084020\",\n",
      "            \"creator\": \"Ragni\",\n",
      "            \"title\": \"Mobile gibberellin directly stimulates Arabidopsis hypocotyl xylem expansion\",\n",
      "            \"publicationYear\": \"2011\",\n",
      "            \"volume\": \"23\",\n",
      "            \"firstPage\": \"1322\",\n",
      "            \"containerTitle\": \"Plant Cell\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib22\",\n",
      "            \"doi\": \"https://doi.org/10.5061/dryad.b835k\",\n",
      "            \"creator\": \"Sankar\",\n",
      "            \"title\": \"Data from: Automated quantitative histology reveals vascular morphodynamics during Arabidopsis hypocotyl secondary growth\",\n",
      "            \"publicationYear\": \"2014\",\n",
      "            \"containerTitle\": \"Dryad Digital Repository\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib23\",\n",
      "            \"doi\": \"https://doi.org/10.1016/j.cub.2008.02.070\",\n",
      "            \"creator\": \"Sibout\",\n",
      "            \"title\": \"Flowering as a condition for xylem expansion in Arabidopsis hypocotyl and root\",\n",
      "            \"publicationYear\": \"2008\",\n",
      "            \"volume\": \"18\",\n",
      "            \"firstPage\": \"458\",\n",
      "            \"containerTitle\": \"Current Biology\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib24\",\n",
      "            \"doi\": \"https://doi.org/10.1111/j.1469-8137.2010.03236.x\",\n",
      "            \"creator\": \"Spicer\",\n",
      "            \"title\": \"Evolution of development of vascular cambia and secondary growth\",\n",
      "            \"publicationYear\": \"2010\",\n",
      "            \"volume\": \"186\",\n",
      "            \"firstPage\": \"577\",\n",
      "            \"containerTitle\": \"The New Phytologist\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib25\",\n",
      "            \"doi\": \"https://doi.org/10.1007/s00138-011-0345-9\",\n",
      "            \"creator\": \"Theriault\",\n",
      "            \"title\": \"Cell morphology classification and clutter mitigation in phase-contrast microscopy images using machine learning\",\n",
      "            \"publicationYear\": \"2012\",\n",
      "            \"volume\": \"23\",\n",
      "            \"firstPage\": \"659\",\n",
      "            \"containerTitle\": \"Machine Vision and Applications\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib26\",\n",
      "            \"doi\": \"https://doi.org/10.1016/j.cell.2012.02.048\",\n",
      "            \"creator\": \"Uyttewaal\",\n",
      "            \"title\": \"Mechanical stress acts via katanin to amplify differences in growth rate between adjacent cells in Arabidopsis\",\n",
      "            \"publicationYear\": \"2012\",\n",
      "            \"volume\": \"149\",\n",
      "            \"firstPage\": \"439\",\n",
      "            \"containerTitle\": \"Cell\"\n",
      "        },\n",
      "        {\n",
      "            \"key\": \"bib27\",\n",
      "            \"doi\": \"https://doi.org/10.1038/ncb2764\",\n",
      "            \"creator\": \"Yin\",\n",
      "            \"title\": \"A screen for morphological complexity identifies regulators of switch-like transitions between discrete cell shapes\",\n",
      "            \"publicationYear\": \"2013\",\n",
      "            \"volume\": \"15\",\n",
      "            \"firstPage\": \"860\",\n",
      "            \"containerTitle\": \"Nature Cell Biology\"\n",
      "        }\n",
      "    ],\n",
      "    \"license\": {\n",
      "        \"id\": \"CC-BY-3.0\",\n",
      "        \"url\": \"https://creativecommons.org/licenses/by/3.0/legalcode\"\n",
      "    },\n",
      "    \"descriptions\": [\n",
      "        {\n",
      "            \"description\": \"Among various advantages, their small size makes model organisms preferred subjects of investigation. Yet, even in model systems detailed analysis of numerous developmental processes at cellular level is severely hampered by their scale. For instance, secondary growth of Arabidopsis hypocotyls creates a radial pattern of highly specialized tissues that comprises several thousand cells starting from a few dozen. This dynamic process is difficult to follow because of its scale and because it can only be investigated invasively, precluding comprehensive understanding of the cell proliferation, differentiation, and patterning events involved. To overcome such limitation, we established an automated quantitative histology approach. We acquired hypocotyl cross-sections from tiled high-resolution images and extracted their information content using custom high-throughput image processing and segmentation. Coupled with automated cell type recognition through machine learning, we could establish a cellular resolution atlas that reveals vascular morphodynamics during secondary growth, for example equidistant phloem pole formation.\",\n",
      "            \"descriptionType\": \"Abstract\"\n",
      "        }\n",
      "    ],\n",
      "    \"fundingReferences\": [\n",
      "        {\n",
      "            \"funderName\": \"SystemsX\"\n",
      "        },\n",
      "        {\n",
      "            \"funderName\": \"EMBO longterm post-doctoral fellowships\"\n",
      "        },\n",
      "        {\n",
      "            \"funderName\": \"Marie Heim-Voegtlin\"\n",
      "        },\n",
      "        {\n",
      "            \"funderName\": \"University of Lausanne\",\n",
      "            \"funderIdentifier\": \"https://doi.org/10.13039/501100006390\",\n",
      "            \"funderIdentifierType\": \"Crossref Funder ID\"\n",
      "        },\n",
      "        {\n",
      "            \"funderName\": \"SystemsX\"\n",
      "        },\n",
      "        {\n",
      "            \"funderName\": \"EMBO\",\n",
      "            \"funderIdentifier\": \"https://doi.org/10.13039/501100003043\",\n",
      "            \"funderIdentifierType\": \"Crossref Funder ID\"\n",
      "        },\n",
      "        {\n",
      "            \"funderName\": \"Swiss National Science Foundation\",\n",
      "            \"funderIdentifier\": \"https://doi.org/10.13039/501100001711\",\n",
      "            \"funderIdentifierType\": \"Crossref Funder ID\"\n",
      "        },\n",
      "        {\n",
      "            \"funderName\": \"University of Lausanne\",\n",
      "            \"funderIdentifier\": \"https://doi.org/10.13039/501100006390\",\n",
      "            \"funderIdentifierType\": \"Crossref Funder ID\"\n",
      "        }\n",
      "    ],\n",
      "    \"provider\": \"Crossref\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "commonmeta = metadata.write()\n",
    "print(commonmeta)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.6"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "a665b5d41d17b532ea9890333293a1b812fa0b73c9c25c950b3cedf1bebd0438"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
